adapters/repos/db/shard_proxy.go

```diff
@@ -0,0 +1,456 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/weaviate/weaviate/entities/multi"
+	"github.com/weaviate/weaviate/entities/searchparams"
+
+	"github.com/weaviate/weaviate/adapters/repos/db/indexcheckpoint"
+	"github.com/weaviate/weaviate/adapters/repos/db/indexcounter"
+	"github.com/weaviate/weaviate/adapters/repos/db/inverted"
+	"github.com/weaviate/weaviate/adapters/repos/db/lsmkv"
+	"github.com/weaviate/weaviate/entities/additional"
+	"github.com/weaviate/weaviate/entities/aggregation"
+	"github.com/weaviate/weaviate/entities/backup"
+	"github.com/weaviate/weaviate/entities/filters"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/schema"
+	"github.com/weaviate/weaviate/entities/search"
+	"github.com/weaviate/weaviate/entities/storagestate"
+	"github.com/weaviate/weaviate/usecases/monitoring"
+	"github.com/weaviate/weaviate/usecases/replica"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/weaviate/weaviate/entities/storobj"
+
+	"github.com/weaviate/weaviate/usecases/objects"
+)
+
+type LazyLoadShard struct {
+	shardOpts *deferredShardOpts
+	shard     *Shard
+	loaded    bool
+	mutex     sync.Mutex
+}
+
+func NewLazyLoadShard(ctx context.Context, promMetrics *monitoring.PrometheusMetrics,
+	shardName string, index *Index, class *models.Class, jobQueueCh chan job,
+	indexCheckpoints *indexcheckpoint.Checkpoints,
+) (*LazyLoadShard, error) {
+	l := &LazyLoadShard{
+		shardOpts: &deferredShardOpts{
+			promMetrics: promMetrics,
+
+			name:             shardName,
+			index:            index,
+			class:            class,
+			jobQueueCh:       jobQueueCh,
+			indexCheckpoints: indexCheckpoints,
+		},
+	}
+	return l, nil
+}
+
+type deferredShardOpts struct {
+	promMetrics      *monitoring.PrometheusMetrics
+	name             string
+	index            *Index
+	class            *models.Class
+	jobQueueCh       chan job
+	indexCheckpoints *indexcheckpoint.Checkpoints
+}
+
+func (l *LazyLoadShard) Load() error {
+	if l.loaded {
+		return nil
+	}
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if l.loaded {
+		return nil
+	}
+	shard, err := NewShard(context.Background(), l.shardOpts.promMetrics, l.shardOpts.name, l.shardOpts.index, l.shardOpts.class, l.shardOpts.jobQueueCh, l.shardOpts.indexCheckpoints)
```

Comment:
I think `Load()` should accept a context, and pass it to `NewShard`, rather than passing in a new context instance. Some of the shard methods (esp in replication use case) use a deadlined context.

Also, why return error if the method only panics?

Reply:
Ooops, pressed the wrong button there.  Sorry everyone!

My other concern is that a context could only be useful during weaviate shutdown, because half-loading a shard during normal weaviate operation is not acceptable.  The load can occur on any method call in Shard, which means that there could be complicated state changes in progress.  It's really succeed-or-shutdown for shard loading, we can't back out halfway.  Except with a panic 😄

Reply:
I agree with @donomii that we cannot have the shard in a halfway state; it is either loaded or not.
However, we can still pass the context. If the context expires, we return an error to the caller, but behind the scenes, we continue loading the shard until completion

Reply:
Ok, we can handle context passing as a later improvement if needed.

Reply:
That's a possibility, but the context would never trigger because most calls to Shard don't have a context.

e.g. Calling shard.Name() can now trigger loading the shard.  We can't reasonably add a context to every trivial accessor that returns a string, and we definitely can't return an error because our code will be a complete mess if we have to do this:

`name, err := shard.Name()
 if err != nil {
    return error("Unable to get shard name")
}

log.Printf("Doing something to shard %s¥n", name)
`